### ISSUE TITLE
no-useless-escape false positive on step definition

### ISSUE THREAD

**Issue by malixsys** [eslint/eslint#7767]
**Tell us about your environment**
    node 6.9.1
    "eslint": "3.12.0",
    "eslint-config-airbnb": "13.0.0",
    "eslint-import-resolver-babel-module": "^2.0.1",
    "eslint-plugin-import": "2.2.0",
    "eslint-plugin-jsx-a11y": "2.2.3",
    "eslint-plugin-react": "6.8.0",

**What parser (default, Babel-ESLint, etc.) are you using?**
    default

**Please show your full configuration:**

```yml
root: true

extends:
    - airbnb

env:
  browser: true

rules:
  comma-dangle: 'off'
  indent: ['error', 4, { SwitchCase: 1, VariableDeclarator: 1, outerIIFEBody: 1 }]
  max-len: ['error', 120, 4, {
    ignoreUrls: true,
    ignoreComments: false,
    ignoreRegExpLiterals: true,
    ignoreStrings: true,
    ignoreTemplateLiterals: true,
  }]

  react/jsx-filename-extension: ['error', { extensions: ['.js', '.jsx'] }]
  react/jsx-indent: ['error', 4]
  react/jsx-indent-props: ['error', 4]
  react/jsx-space-before-closing: ['error', never]
  react/prop-types: ['error', { ignore: [children] }]
  # A little bit buggy for now... See https://github.com/yannickcr/eslint-plugin-react/issues/855
  react/no-unused-prop-types: ['off', {
    customValidators: [],
    skipShapeProps: true,
  }]
  # A lot of components use the 'array' and 'object' proptype, so disabling it for now
  react/forbid-prop-types: 'off'

  import/no-extraneous-dependencies: ['error', { devDependencies: true }]
  import/no-named-as-default: 'off'

  # A lot of our components and UIF components are not really a11y ready...
  jsx-a11y/anchor-has-content: 'off'
  jsx-a11y/no-static-element-interactions: 'off'

settings:
  import/resolver:
    babel-module: {}
```
**What did you do? Please include the actual source code causing the issue.**

I have cucumber.js step definition and one gets flagged with no-useless-escape 
```js
// works but gets flagged
    this.When(/^I type \[(\S*)\] in the editor$/, (client, key) =>
        client.page.editor().typeKey(key)
    );
// this no-useless-escape passing regex doesn't work, step not found 
    this.When(/^I type [(\S*)] in the editor$/, (client, key) =>
        client.page.editor().typeKey(key)
    );
```

**What did you expect to happen?**
/^I type \[(\S*)\] in the editor$/ should not get flagged

**What actually happened? Please include the actual, raw output from ESLint.**
```editor_steps.js: Unnecessary escape character: \]. (no-useless-escape)```

**Comment by platinumazure** [comment]
As the rule notes, `\]` is in fact a useless escape when it's not inside a character class.

What the rule wants you to do (to get the literal brackets to show up) is this:

```js
    this.When(/^I type \[(\S*)] in the editor$/, (client, key) =>
        client.page.editor().typeKey(key)
    );
```

Note that the opening bracket *is* escaped (to avoid creating a character class), but the closing bracket is *not* escaped.

**Comment by malixsys** [comment]
Ah OK! Excellent! The `--fix` was over-zealous then...

**Comment by platinumazure** [comment]
@malixsys Please tell me more, we really want our autofixer not to break working code...

**Comment by not-an-aardvark** [comment]
I'm also confused, because `no-useless-escape` doesn't have an autofixer.

**Comment by malixsys** [comment]
Nevermind... My bad. @platinumazure
